keps/sig-scheduling/20190926-default-even-pod-spreading.md

```diff
@@ -0,0 +1,294 @@
+
+## Motivation
+
+In order for a workload (pod) to use `EvenPodsSpreadPriority`:
+
+1. Authors have to have an idea of the underlying topology.
+1. PodSpecs become less portable if their spreading constraints are tailored to a specific topology.
+
+On the other hand, cluster operators know the underlying topology of the cluster, which makes
+them suitable to provide default spreading constraints for all workloads in their cluster.
+
+### Goals
+
+- Cluster operators can define default spreading constraints for pods that don't provide any
+  `pod.spec.topologySpreadConstraints`.
+- Workloads are spread with the default constraints if they belong to the same service, replication controller,
+  replica set or stateful set, and if they don't define `pod.spec.topologySpreadConstraints`.
+- Provide a k8s default for `topologySpreadConstraints` that produces a priority equivalent to
+  `SelectorSpreadPriority`, so that this algorithm can be removed from the default algorithms' provider.
+
+### Non-Goals
+
+- Set defaults for specific namespaces or according to other selectors.
+- Removal of `ServiceSpreadingPriority` or `ServiceAntiAffinity` priorities.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a cluster operator, I want to set default spreading constraints for workloads in the cluster.
+Currently, `SelectorSpreadPriority` provides a canned priority that spreads across nodes
+and zones (`failure-domain.beta.kubernetes.io/zone`). However, the nodes in my cluster have
+custom topology keys (for physical host, rack, etc.).
+
+#### Story 2
+
+As a workload author, I want to spread the workload in the cluster, but:
+(1) I don't know the topology of the cluster I'm running on.
+(2) I want to be able to run my PodSpec in different clusters (on-prem and cloud).
+
+### Implementation Details/Notes/Constraints
+
+
+#### Relationship with SelectorSpreadingPriority
+
+Note that Default `topologySpreadConstraints` has a similar effect to `SelectorSpreadingPriority`.
+Given that the latter is not configurable, they could return conflicting priorities, which
+may not be the intention of an operator. On the other hand, a proper Default for `topologySpreadConstraints`
+could provide the same priority as `SelectorSpreadingPriority`. Thus, there's no need for the
+features to co-exist.
+
+Give that we guard Default `topologySpreadConstraints` behind a feature flag,
+these would be its semantics:
+
+- If the feature is enabled, `SelectorSpreadingPriority` is removed from the default set of priorities.
+  K8s provides the Default `topologySpreadConstraints` that matches the priority given by
+  `SelectorSpreading` if the cluster operator doesn't specify one.
+- If the cluster operator provides a Policy that includes `SelectorSpreadingPriority` and
+  `EvenPodsSpreadPriority`, K8s provides an empty Default `topologySpreadConstraints`.
+  The cluster operator can still specify Default `topologySpreadConstraints`,
+  in which case both priorities run.
```

Comment:
I don't we should run both priorities, I think we should clarify that if both are specified, then one overrides the other.

Reply:
So your take would be that `EvenPodsSpreadPriority` overrides and disables `SelectorSpreadingPriority`?

Reply:
We should have a single implementation, a plugin. This plugin will either be configured using the priority-to-plugin translation logic we have if ```SelectorSpreadingPriority``` is specified, otherwise it is configured using the new ComponentConfig parameters you will be adding.

Reply:
See updated clarifications